pkg/destroy/openstack/openstack.go

```diff
@@ -347,49 +347,78 @@ func deletePorts(opts *clientconfig.ClientOpts, filter Filter, logger logrus.Fie
-					logger.Errorf("While deleting port %q, the update of the floating IP %q failed with error: %v", port.ID, fip.ID, err)
-					continue
+
+	// Prefetch list of FIPs to save list calls for each port
+	allPages, err = floatingips.List(conn, floatingips.ListOpts{}).AllPages()
+	if err != nil {
+		logger.Error(err)
+		return false, nil
+	}
+	allFIPs, err := floatingips.ExtractFloatingIPs(allPages)
+	if err != nil {
+		logger.Error(err)
+		return false, nil
+	}
+
+	// Organize FIPs for easy lookup
+	fipByPort := make(map[string]floatingips.FloatingIP)
+	for _, fip := range allFIPs {
+		fipByPort[fip.PortID] = fip
+	}
+
+	deletePortsChunk := func(portsChunk []ports.Port, deletedChannel chan int) {
+		localDeleted := 0
+		for _, port := range portsChunk {
+			// If a user provisioned floating ip was used, it needs to be dissociated.
+			// Any floating Ip's associated with ports that are going to be deleted will be dissociated.
+			if fip, ok := fipByPort[port.ID]; ok {
+				logger.Debugf("Dissociating Floating IP %q", fip.ID)
+				_, err := floatingips.Update(conn, fip.ID, floatingips.UpdateOpts{}).Extract()
+				if err != nil {
+					// Ignore the error if the floating ip cannot be found and return with an appropriate message if it's another type of error
+					var gerr gophercloud.ErrDefault404
+					if !errors.As(err, &gerr) {
+						// Just log the error and move on to the next port
+						logger.Errorf("While deleting port %q, the update of the floating IP %q failed with error: %v", port.ID, fip.ID, err)
+						continue
+					}
+					logger.Debugf("Cannot find floating ip %q. It's probably already been deleted.", fip.ID)
 				}
-				logger.Debugf("Cannot find floating ip %q. It's probably already been deleted.", fip.ID)
 			}
-		}
 
-		logger.Debugf("Deleting Port %q", port.ID)
-		err = ports.Delete(conn, port.ID).ExtractErr()
-		if err != nil {
-			// This can fail when port is still in use so return/retry
-			// Just log the error and move on to the next port
-			logger.Debugf("Deleting Port %q failed with error: %v", port.ID, err)
-			// Try to delete associated trunk
-			deleteAssociatedTrunk(conn, logger, port.ID)
-			continue
+			logger.Debugf("Deleting Port %q", port.ID)
+			err = ports.Delete(conn, port.ID).ExtractErr()
+			if err != nil {
+				// This can fail when port is still in use so return/retry
+				// Just log the error and move on to the next port
+				logger.Debugf("Deleting Port %q failed with error: %v", port.ID, err)
+				// Try to delete associated trunk
+				deleteAssociatedTrunk(conn, logger, port.ID)
+				continue
+			}
+			localDeleted++
 		}
-		numberDeleted++
+		deletedChannel <- localDeleted
+	}
+
+	deletedChannel := make(chan int)
+	const chunksNumber = 10
+
+	// run processing in chunks
+	for i := 0; i < chunksNumber; i++ {
+		// This is at trick, but it'll produce chunksNumber chunks of roughly the same size.
+		// It may produce empty chunks, but why would we care?
```

Comment:
This doesn't seem right. Take for example `len(allPorts) = 1`, you'd have 10 workers trying to delete the same port, and in the end `numberDeleted` is equal to 10.

You should instead use a buffered channel and run each port delete in its own goroutine.
See https://gobyexample.com/worker-pools for example.

Reply:
This is actually fun, in such case the trick in 411-412 would produce 10 chunks, 9 of which would be empty. Check out and play with [1]. It's suboptimal but I opted to do it that way to make code cleaner.

Now the way you're proposing seems to be even simpler, let me try it out.

[1] https://go.dev/play/p/ogvz3ARDGc9

Reply:
Beautiful!
I used this model to test your code: https://go.dev/play/p/6h2w0YgXts3

Reply:
Right, you've tested the way @mandre proposed, which can also be called "the correct way of doing this". ;)